server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -514,6 +514,10 @@ private void reroute(RoutingAllocation allocation, Consumer<RoutingAllocation> r
 
         removeDelayMarkers(allocation);
 
+        if (isBalancedShardsAllocator()) {
+            allocation.getGatewayAllocations(); // no need to track gateway allocations for the legacy allocator
```

Comment:
NIT: consider renaming to `getAndResetGatewayAllocations()` otherwise it looks like this is not doing anything